### DISCUSSION TITLE
Clarity on Iris Core Developer Criterion and Nomination

### DISCUSSION THREAD

**Discussion by bjlittle** [SciTools/iris#5239] [Ideas]
One thing that's always slightly irked me, is our lack of clarity to the community (and ourselves) on how a **contributor** can become an `iris` **core developer**.

For me, I'd like to have agreement on (at least) the following open questions:

1. What is the role and responsibilities of an `iris` core developer?
2. What is the minimum criterion that a contributor has to satisfy?
3. Nomination:
   a. How can a contributor make existing `iris` core developers aware that they want to become a core developer?
   b. How do existing `iris` core developers nominate a specific contributor? 
4. How do existing `iris` core developers discuss a nomination?
5. How do we communicate the outcome of the nomination to the contributor?
6. How do we publicise to the community our awesome, new `iris` core developer?
7. Where do we capture and share our agreement to all of the above?

Ideally, I'd hope that the outcome of this discussion is some flexible and light-weight recommendations that we can evolve over time. 

At the end of the day, having _something concrete_ in place will be so much better than the mysterious randomness that we have at the moment.

**Comment by bjlittle** [comment]
I'll start the ball rolling...

**1. What is the role and responsibilities of an iris core developer?**

For me, a developer that has spare cycles to:
1. Be a good citizen, role model and ambassador within the community. Hopefully the outcome of this [poll](https://github.com/SciTools/iris/discussions/5197) will encapsulate some of the behaviours/values that we measure ourselves by
2. Raises issues to highlight changes "for the better" within `iris`
3. Comments on issues and pull-requests, giving value where they can
4. Reviews and/or merges open pull-requests
5. Nominates other contributors as core developers
6. Recommends best practice

**Comment by bjlittle** [comment]
**2. What is the minimum criterion that a contributor has to satisfy?**

Takes a sustained active interest in `iris`, which is demonstrated by the following:

- Commented on assorted issues and/or pull-requests
- Participated in our community GitHub Discussions 
- Contributed several valued pull-requests to the core of `iris`
- Regarded as a "safe-pair-of-hands" :tm: (@rcomer)

**Reply by rcomer** [reply]
A safe pair of hands.  If someone has merge rights, we want to know that they are going to be careful about what they merge!  I'm not sure how you measure that, but if someone has left some reviews it probably gives some indication about how they think about additions to the project.

**Reply by trexfeathers** [reply]
We have in the past used core developer status before the above points were demonstrated, as a means to _encourage_ the above points. We'd need some other form of encouragement instead.

**Reply by bjlittle** [reply]
> We have in the past used core developer status before the above points were demonstrated, as a means to _encourage_ the above points. We'd need some other form of encouragement instead.

@trexfeathers We do have the `triage` teams with limited role perms that allows people to get more involved... is this the kinda thing that you're thinking about?

For me, `triage` is an ideal stepping stone (and proving ground) to greater responsibility.

**Reply by trexfeathers** [reply]
Yes, I am thinking of some older occasions and perhaps if we had `triage` back then we would have used it.

**Comment by bjlittle** [comment]
**3a. How can a contributor make existing iris core developers aware that they want to become a core developer?**

I'd suggest either raising an issue or opening a GH Discussion (GHD). 

Does anyone have any better ideas? Should be have a dedicated issue or GHD that contributors tag us on?

**Reply by bjlittle** [reply]
If the steps to take are clear to the community on how they become a core developer, should contributors nominate themselves when they think they've ticked all the boxes?

**Reply by rcomer** [reply]
As with 3b, I think a private method should be available, though I'm not sure whether contributors would or should nominate themselves...

If someone thinks "I want to help this project by leaving reviews and comments" then that's great and they can do that without any elevated permissions.  If they do a good job of it then it makes sense to give them the power to merge PRs, etc.  Maybe we could make it more obvious that anyone is welcome to leave reviews.  I appreciated [this tiny section of the Matplotlib dev guide](https://matplotlib.org/stable/devel/triage.html#working-on-prs-to-help-review), as without it I wouldn't have wanted to comment on someone else's PR for fear of stepping on core devs' toes.

**Comment by bjlittle** [comment]
**3b. How do existing iris core developers nominate a specific contributor?**

Ideally, I'd love for this process to be as open and transparent as possible.

Perhaps an `iris` core dev could use a GHD to make the nomination, citing the evidence to support their proposal. 

If 3 or more core devs :+1: the GHD, then we consider the nomination as successful.

Alternatively, if we want to nominate/discuss this matter in private (as contributors may not want to see such discussions unfold in public) @lbdreyer suggested that the nomination and subsequent discussion could happen on GitHub in private i.e., conversations only visible to other core devs

**Reply by rcomer** [reply]
I think it should probably be private: if someone is public nominated but not ultimately invited then that could be embarrassing for the contributor in question.

**Reply by trexfeathers** [reply]
https://github.com/SciTools/iris/discussions/5239#discussioncomment-5591493

**Comment by bjlittle** [comment]
**4. How do existing iris core developers discuss a nomination?**

I guess this depends on the outcome of **3b** ? :thinking:

**Reply by trexfeathers** [reply]
I would prefer the conversation take place in private. However, with [the retirement of Team Discussions](https://github.blog/changelog/2023-02-08-sunset-notice-team-discussions/), private conversations will no longer be possible within GitHub.

We could conceivably create a dedicated repository for this, with locked down permissions, but even that wouldn't offer any granularity for different groups of people.

**Reply by rcomer** [reply]
Private rooms can be created on Gitter, and you can log in with your GitHub account...

**Reply by trexfeathers** [reply]
A bit hacky, but it is also possible to create private projects and invite collaborators. You can invite a whole team, e.g. `iris-devs` (I just tested this). So we could make short-lived projects purely for proposing/discussing something like this, then close once resolved.

**Comment by bjlittle** [comment]
**5. How do we communicate the outcome of the nomination to the contributor?**

The contributor must have signed the CLA, therefore we have their contact details.

It would be great if we can celebrate a successful nomination publicly, but I guess we should get prior agreement from the contributor.

**Comment by bjlittle** [comment]
**6. How do we publicise to the community our awesome, new iris core developer?**

At the very least, make a song and dance, blow the trumpet and go to town in the `whatsnew` of the next release.

Also, would we want another dedicated place to celebrate this? Perhaps on Twitter?

**Reply by rcomer** [reply]
Twitter and whatsnew seem good to me.

Edit: though on Twitter probably don’t tag them without checking they are happy with that first.  They might like to keep Twitter life and GitHub life separate.

**Reply by trexfeathers** [reply]
A cute touch that I like: the new developer should be the one to merge the What's New addition.

**Comment by bjlittle** [comment]
**7. Where do we capture and share our agreement to all of the above?**

I'd suggest an ideal place might be a dedicated section in the [Developers Guide](https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_getting_involved.html#development-where-to-start) in our documentation.
